### PR TITLE
(WIP) issue/6513 - Date/Time/Format/Zone improvements:

### PR DESCRIPTION
**Pull request by JJJ**
Work in Progress.

See #6513.

### CODE REVIEW

File: includes/admin/discounts/add-discount.php [hunk 1]
```diff
@@ -141,8 +141,8 @@
 					<label for="edd-start"><?php _e( 'Start date', 'easy-digital-downloads' ); ?></label>
 				</th>
 				<td>
-					<input name="start_date" id="edd-start" type="text" value="" class="edd_datepicker"/>
-					<p class="description"><?php _e( 'Enter the start date for this discount code in the format of mm/dd/yyyy. For no start date, leave blank. If entered, the discount can only be used after or on this date.', 'easy-digital-downloads' ); ?></p>
+					<input name="start_date" id="edd-start" type="text" value="" class="edd_datepicker" data-format="<?php echo edd_get_date_picker_format(); ?>" placeholder="<?php echo edd_get_date_picker_format(); ?>" />
```
Comment (DrewAPicture) [inline includes/admin/discounts/add-discount.php]: Escape?

File: includes/admin/payments/view-order-details.php [hunk 2]
```diff
@@ -105,7 +105,7 @@
 										<div class="edd-admin-box-inside">
 											<p>
 												<span class="label"><?php _e( 'Date:', 'easy-digital-downloads' ); ?></span>&nbsp;
-												<input type="text" name="edd-payment-date" value="<?php echo esc_attr( date( 'm/d/Y', $payment_date ) ); ?>" class="medium-text edd_datepicker"/>
+												<input type="text" name="edd-payment-date" value="<?php echo esc_attr( date( 'Y-m-d', $payment_date ) ); ?>" class="medium-text edd_datepicker" placeholder="<?php echo edd_get_date_picker_format(); ?>" />
```
Comment (DrewAPicture) [inline includes/admin/payments/view-order-details.php]: Why the change in entered format? The datepicker reformats to Y-m-d on `$_POST` anyway, doesn't it?

File: includes/admin/reporting/graphing.php [hunk 3]
```diff
@@ -667,11 +667,19 @@ function edd_reports_graph_controls() {
...
+								'name'        => 'filter_from',
+								'placeholder' => 'yyyy-mm-dd',
+								'value'       => $from
+							) ); ?>
 
 							<span><?php _ex( 'To', 'date filter', 'easy-digital-downloads' ); ?></span>
-							<?php echo EDD()->html->date_field( array( 'name' => 'filter_to', 'placeholder' => __( 'mm/dd/yyyy', 'easy-digital-downloads' ), 'value' => $to ) ); ?>
+							<?php echo EDD()->html->date_field( array(
+								'name'        => 'filter_to',
+								'placeholder' => 'yyyy-mm-dd',
```
Comment (DrewAPicture) [inline includes/admin/reporting/graphing.php]: Same question on the entered format. I think it was previously mm/dd/yyyy because that's "more familiar".

File: includes/class-edd-utilities.php [hunk 4]
```diff
@@ -126,44 +134,65 @@ public function get_registry( $name ) {
...
 
-		switch( $format ) {
-			case 'time':
-				$format = $this->get_time_format();
+		// What known format are we getting?
+		switch ( $format ) {
+
+			// jQuery UI Datepicker fields, placeholders, etc...
+			case 'date-attribute' :
+				$retval = 'yyyy-mm-dd';
```
Comment (DrewAPicture) [inline includes/class-edd-utilities.php]: Purely a style question, but ... why the variable name change from something self-documenting (`$format`) to something that's indicative only of being a return value?

File: includes/date-functions.php [hunk 5]
```diff
@@ -80,6 +55,48 @@ function edd_get_timezone() {
  *
  * @return string date_format()-compatible date format string.
  */
-function edd_get_date_format( $format ) {
+function edd_get_date_format( $format = 'date' ) {
```
Comment (DrewAPicture) [inline includes/date-functions.php]: Documentation for `$format`?

File: includes/date-functions.php [hunk 6]
```diff
@@ -80,6 +55,48 @@ function edd_get_timezone() {
...
+ */
+function edd_get_date_picker_format( $context = 'display' ) {
+
+	// What is the context that we are getting the picker format for?
+	switch ( $context ) {
+
+		// jQuery UI Datepicker does its own thing
+		case 'js' :
+		case 'javascript' :
+			$retval = EDD()->utils->get_date_format_string( 'date-js' );
```
Comment (DrewAPicture) [inline includes/date-functions.php]: Same question on `$retval` vs something self-documenting like `$format`.

File: includes/date-functions.php [hunk 7]
```diff
@@ -80,6 +55,48 @@ function edd_get_timezone() {
...
+
+		// Used to display in an attribute, placeholder, etc...
+		case 'display' :
+		default :
+			$retval = EDD()->utils->get_date_format_string( 'date-attribute' );
+			break;
+	}
+
+	// Filter & return
+	return apply_filters( 'edd_get_date_picker_format', $retval, $context );
```
Comment (DrewAPicture) [inline includes/date-functions.php]: Missing hook doc.

File: includes/emails/class-edd-email-tags.php [hunk 8]
```diff
@@ -833,11 +833,9 @@ function edd_email_tag_payment_method( $payment_id ) {
  * Email template tag: sitename
  * Your site name
  *
- * @param int $payment_id
- *
  * @return string sitename
  */
-function edd_email_tag_sitename( $payment_id ) {
```
Comment (DrewAPicture) [inline includes/emails/class-edd-email-tags.php]: If we're removing a parameter, we need to properly deprecate it.

Comment (JJJ) [reply includes/emails/class-edd-email-tags.php]: This parameter should never have been here. It never did anything, and was never used.

File: includes/date-functions.php (hunk 6 above)
Comment (JJJ) [reply includes/date-functions.php]: Because it’s a return value based on the context.

Naming it something like `$retval` informs developers that that’s the variable that will inevitably be returned by the function.

Naming it `$format` does not indicate its the return value. It also possibly collides with other local variables or parameters.

`$retval` is like a reserved keyword, IMO. Use it only to define the variable used to filter and return something.

File: includes/admin/reporting/graphing.php (hunk 3 above)
Comment (JJJ) [reply includes/admin/reporting/graphing.php]: The jQuery UI Datepicker got changed to `yy-mm-dd` so it no longer matches the format or pattern.

I did forget to use the new function here though, so I’ll note that.

File: includes/admin/payments/view-order-details.php (hunk 2 above)
Comment (JJJ) [reply includes/admin/payments/view-order-details.php]: This is so that the date matches the placeholder text and the returned format of jQuery UI Datepicker.

File: includes/admin/discounts/add-discount.php (hunk 1 above)
Comment (JJJ) [reply includes/admin/discounts/add-discount.php]: Good call.

I’ll probably end up with a new function to handle the escape and echo.

File: includes/date-functions.php (hunk 7 above)
Comment (JJJ) [reply includes/date-functions.php]: Added!

File: includes/date-functions.php (hunk 5 above)
Comment (JJJ) [reply includes/date-functions.php]: Done!

File: includes/admin/reporting/graphing.php (hunk 3 above)
Comment (JJJ) [reply includes/admin/reporting/graphing.php]: Updated to use `edd_get_date_picker_format()` instead of `yyyy-mm-dd` or any other hard-coded string.

File: includes/admin/discounts/add-discount.php (hunk 1 above)
Comment (JJJ) [reply includes/admin/discounts/add-discount.php]: Done!

**Comment by JJJ** [comment]
These changes here are not very huge, and the code itself is OK and purposeful, so I just merged it.

Can continue to iterate on this if/as needed.
